File: servers/typescript/src/handler_system.ts
import type { StarfishFrame } from "./types.js";
import type { Client } from "./client.js";
import type { StarfishServer } from "./starfish_server.js";
import { parseTo } from "./types.js";
import { createErrorFrame, ERR_PROTOCOL_INVALID_FRAME } from "./errors.js";

export function handleClockSync(
  hub: StarfishServer,
  client: Client,
  frame: StarfishFrame,
): void {
  const now = Date.now();
  client.sendFrame({
    header: {
      id: hub.idGen.messageId(),
      resource: "clock",
      method: "sync",
      kind: "response",
      replyTo: frame.header.id,
    },
    payload: { status: "ok", serverTime: now },
  });
}

export function handleAck(
  hub: StarfishServer,
  client: Client,
  frame: StarfishFrame,
): void {
  routeReply(hub, client, frame);
}

export function handleNack(
  hub: StarfishServer,
  client: Client,
  frame: StarfishFrame,
): void {
  routeReply(hub, client, frame);
}

function routeReply(hub: StarfishServer, client: Client, frame: StarfishFrame): void {
  if (!frame.header.replyTo) {
    client.sendFrame(
      createErrorFrame(hub.idGen, frame.header.id, ERR_PROTOCOL_INVALID_FRAME, frame.header.resource, frame.header.method),
    );
    return;
  }

  frame.header.from = client.id;
  const targets = parseTo(frame.header.to);
  if (targets.length === 0) return;

  for (const targetId of targets) {
    const target = hub.getClient(targetId);
    if (target) target.sendFrame(frame);
  }
}

File: servers/python/src/starfish_server/handler_system.py
from __future__ import annotations

import time
from typing import TYPE_CHECKING

from .client import Client
from .errors import ERR_PROTOCOL_INVALID_FRAME, create_error_frame
from .types import StarfishFrame, parse_to

if TYPE_CHECKING:
    from .server import StarfishServer


def handle_clock_sync(hub: StarfishServer, client: Client, frame: StarfishFrame) -> None:
    header = frame.get("header", {})
    now = int(time.time() * 1000)
    client.send_frame({
        "header": {
            "id": hub.id_gen.message_id(),
            "resource": "clock",
            "method": "sync",
            "kind": "response",
            "replyTo": header.get("id", ""),
        },
        "payload": {"status": "ok", "serverTime": now},
    })


def handle_ack(hub: StarfishServer, client: Client, frame: StarfishFrame) -> None:
    _route_reply(hub, client, frame)


def handle_nack(hub: StarfishServer, client: Client, frame: StarfishFrame) -> None:
    _route_reply(hub, client, frame)


def _route_reply(hub: StarfishServer, client: Client, frame: StarfishFrame) -> None:
    header = frame.get("header", {})

    if not header.get("replyTo"):
        client.send_frame(
            create_error_frame(
                hub.id_gen, header.get("id", ""), ERR_PROTOCOL_INVALID_FRAME, header.get("resource"), header.get("method")
            )
        )
        return

    header["from"] = client.id
    targets = parse_to(header.get("to"))
    if not targets:
        return

    for target_id in targets:
        target = hub.get_client(target_id)
        if target:
            target.send_frame(frame)
